kubetest/azure.go

```diff
@@ -0,0 +1,520 @@
+	}
+
+	if *acsWinBinariesURL != "" {
+		v.Properties.OrchestratorProfile.KubernetesConfig.CustomWindowsPackageURL = *acsWinBinariesURL
+	} else if c.acsCustomWinBinariesURL != "" {
+		v.Properties.OrchestratorProfile.KubernetesConfig.CustomWindowsPackageURL = c.acsCustomWinBinariesURL
+	}
+	apiModel, _ := json.Marshal(v)
+	c.apiModelPath = path.Join(c.outputDir, "kubernetes.json")
+	err := ioutil.WriteFile(c.apiModelPath, apiModel, 0644)
+	if err != nil {
+		return fmt.Errorf("Cannot write to file: %v", err)
+	}
+	return nil
+}
+
+func (c *Cluster) getAcsEngine(retry int) error {
+	downloadPath := path.Join(os.Getenv("HOME"), "acs-engine.tar.gz")
+	f, err := os.Create(downloadPath)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	for i := 0; i < retry; i++ {
+		log.Printf("downloading %v from %v.", downloadPath, *acsEngineURL)
+		if err := httpRead(*acsEngineURL, f); err == nil {
+			break
+		}
+		err = fmt.Errorf("url=%s failed get %v: %v.", *acsEngineURL, downloadPath, err)
+		if i == retry-1 {
+			return err
+		}
+		log.Println(err)
+		sleep(time.Duration(i) * time.Second)
+	}
+
+	f.Close()
+	if *acsEngineMD5 != "" {
+		o, err := control.Output(exec.Command("md5sum", f.Name()))
+		if err != nil {
+			return err
+		}
+		if strings.Split(string(o), " ")[0] != *acsEngineMD5 {
+			return fmt.Errorf("Wrong md5 sum for acs-engine.")
+		}
+	}
+
+	cwd, err := os.Getwd()
+	if err != nil {
+		return fmt.Errorf("Unable to get current directory: %v .", err)
+	}
+	log.Printf("Extracting tar file %v into directory %v .", f.Name(), cwd)
+
+	if err = control.FinishRunning(exec.Command("tar", "-xzf", f.Name(), "--strip", "1")); err != nil {
+		return err
+	}
+	c.acsEngineBinaryPath = path.Join(cwd, "acs-engine")
+	return nil
+
+}
+
+func (c Cluster) generateARMTemplates() error {
+	if err := control.FinishRunning(exec.Command(c.acsEngineBinaryPath, "generate", c.apiModelPath, "--output-directory", c.outputDir)); err != nil {
+		return fmt.Errorf("Failed to generate ARM templates: %v.", err)
+	}
+	return nil
+}
+
+func (c *Cluster) loadARMTemplates() error {
+	var err error
+	template, err := ioutil.ReadFile(path.Join(c.outputDir, "azuredeploy.json"))
+	if err != nil {
+		return fmt.Errorf("Error reading ARM template file: %v.", err)
+	}
+	c.templateJSON = make(map[string]interface{})
+	err = json.Unmarshal(template, &c.templateJSON)
+	if err != nil {
+		return fmt.Errorf("Error unmarshall template %v", err.Error())
+	}
+	parameters, err := ioutil.ReadFile(path.Join(c.outputDir, "azuredeploy.parameters.json"))
+	if err != nil {
+		return fmt.Errorf("Error reading ARM parameters file: %v", err)
+	}
+	c.parametersJSON = make(map[string]interface{})
+	err = json.Unmarshal(parameters, &c.parametersJSON)
+	if err != nil {
+		return fmt.Errorf("Error unmarshall parameters %v", err.Error())
+	}
+	c.parametersJSON = c.parametersJSON["parameters"].(map[string]interface{})
+
+	return nil
+}
+
+func (c *Cluster) getARMClient(ctx context.Context) error {
```

Comment:
can we comment some more of this? I first read this and thought ARM == ARM core not Azure Resource Manager :^)

Reply:
Added comments in latest commit.